tests/nonbonded/test_nonbonded_all_pairs.py

```diff
@@ -0,0 +1,128 @@
+    with pytest.raises(RuntimeError) as e:
+        potential_interp.execute(np.zeros((1, 3)), np.zeros((1, 3)), np.eye(3), 0)
+
+    assert "NonbondedAllPairs::execute_device(): expected P == M*N_*3, got P=3, M*N_*3=6" in str(e)
+
+
+def make_ref_potential(lambda_plane_idxs, lambda_offset_idxs, beta, cutoff):
+    @functools.wraps(nonbonded.nonbonded_v3)
+    def wrapped(conf, params, box, lamb):
+        num_atoms, _ = conf.shape
+        no_rescale = np.ones((num_atoms, num_atoms))
+        return nonbonded.nonbonded_v3(
+            conf,
+            params,
+            box,
+            lamb,
+            charge_rescale_mask=no_rescale,
+            lj_rescale_mask=no_rescale,
+            beta=beta,
+            cutoff=cutoff,
+            lambda_plane_idxs=lambda_plane_idxs,
+            lambda_offset_idxs=lambda_offset_idxs,
+        )
+
+    return wrapped
+
+
+@pytest.mark.parametrize("lamb", [0.0, 0.1])
+@pytest.mark.parametrize("beta", [2.0])
+@pytest.mark.parametrize("cutoff", [1.1])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 1e-4, 5e-4)])
+@pytest.mark.parametrize("num_atoms", [33, 65, 231, 1050, 4080])
+def test_nonbonded_all_pairs_correctness(
+    num_atoms,
+    precision,
+    rtol,
+    atol,
+    cutoff,
+    beta,
+    lamb,
+    example_nonbonded_params,
+    example_conf,
+    example_box,
+    rng: np.random.Generator,
+):
+    "Compares with jax reference implementation."
+
+    conf = example_conf[:num_atoms]
+    params = example_nonbonded_params[:num_atoms, :]
+
+    lambda_plane_idxs = rng.integers(-2, 3, size=(num_atoms,), dtype=np.int32)
+    lambda_offset_idxs = rng.integers(-2, 3, size=(num_atoms,), dtype=np.int32)
+
+    ref_potential = make_ref_potential(lambda_plane_idxs, lambda_offset_idxs, beta, cutoff)
+    test_potential = NonbondedAllPairs(lambda_plane_idxs, lambda_offset_idxs, beta, cutoff)
+
+    GradientTest().compare_forces(
+        conf, params, example_box, lamb, ref_potential, test_potential, precision=precision, rtol=rtol, atol=atol
+    )
+
+
+@pytest.mark.parametrize("lamb", [0.0, 0.1, 0.9, 1.0])
+@pytest.mark.parametrize("beta", [2.0])
+@pytest.mark.parametrize("cutoff", [1.1])
+@pytest.mark.parametrize("precision,rtol,atol", [(np.float64, 1e-8, 1e-8), (np.float32, 2e-4, 5e-4)])
+@pytest.mark.parametrize("num_atoms", [33, 231, 4080])
+def test_nonbonded_all_pairs_interpolated_correctness(
+    num_atoms,
+    precision,
+    rtol,
+    atol,
+    cutoff,
+    beta,
+    lamb,
+    example_nonbonded_params,
+    example_conf,
+    example_box,
+    rng: np.random.Generator,
+):
+    "Compares with jax reference implementation, with parameter interpolation."
+
+    conf = example_conf[:num_atoms]
+    params_initial = example_nonbonded_params[:num_atoms, :]
+    params_final = params_initial + np.where(params_initial, rng.normal(0, 0.01, size=params_initial.shape), 0)
```

Comment:
Conditioning on the initial params being nonzero was necessary to prevent fixed-point overflow for sizes greater than ~2500. I'm not sure I fully understand this, but it seems like this might be the correct thing to do anyway (since LJ param == 0 <==> LJ disabled) and we might want to update the other interpolation tests with this condition

Reply:
>Conditioning on the initial params being nonzero was necessary to prevent fixed-point overflow for sizes greater than ~2500, presumably due to large LJ energies when we turn on LJ interactions that were initially disabled?

That sounds plausible.

--------

Tangent: The workaround `rng.normal(0, 0.01, ...)` can assign negative values to epsilon and sigma -- charges can be negative, but epsilon and sigma should always be non-negative.

Question: Do we have (or want to add) a warning if epsilon or sigma < 0?

Reply:
> The workaround rng.normal(0, 0.01, ...) can assign negative values to epsilon and sigma -- charges can be negative, but epsilon and sigma should always be non-negative.

Makes sense; I hadn't considered this. In d7c4095 I changed the generator to avoid negative values for sigma and epsilon.

> Question: Do we have (or want to add) a warning if epsilon or sigma < 0?

We do not have this currently; it sounds like a good idea to add a check. Unfortunately, adding this didn't seem totally trivial, since it currently requires duplicating the check code in each of the 3 nonbonded potentials. Is it OK with you if I add this as a subtask to #639 ?

Reply:
>In https://github.com/proteneer/timemachine/commit/d7c4095b6da7abfad6abf91416b26c86be8d740f I changed the generator to avoid negative values for sigma and epsilon.

Looks good!

>Unfortunately, adding this didn't seem totally trivial, since it currently requires duplicating the check code in each of the 3 nonbonded potentials. Is it OK with you if I add this as a subtask to https://github.com/proteneer/timemachine/issues/639 ?

Fine by me!

(Non-blocking for the current PR, and probably also not worth doing in a way that introduces much duplication)